BLUESPAWN-client/headers/util/filesystem/FileSystem.h

```diff
@@ -12,13 +10,217 @@
+		short Create();
+
+		/**
+		* Function to delete the file
+		*
+		* @return 1 if deletion was successful, 0 if unsuccessful
+		*/
+		short Delete();
+
+		/**
+		* Function to truncate or extend file length
+		*
+		* @param length - new length of the file in bytes
+		*
+		* @return 1 if trucation or extension was successful, 0 if unsuccessful
+		*/
+		short ChangeFileLength(IN const long length);
+
+		~File();
+	};
 
-bool CheckFileExists(LPCWSTR);
-string GetFileContents(LPCWSTR);
-bool HashFileMD5(LPCWSTR, string&);
+	class Folder {
+		//Path to the current folder
+		LPCWSTR FolderPath;
+		//Whether or not the current folder exists
+		bool FolderExists;
+		//Handle to current file or directory
+		HANDLE hCurFile;
+		//Is the current handle a file or directory
+		bool IsFile;
+		//Information about found files
+		WIN32_FIND_DATA ffd;
+	public:
+		/**
+		* Constructor for the folder object
+		* 
+		* @param path - the path to the folder
+		*/
+		Folder(LPCWSTR path);
+		
+		/**
+		* Function to move to the next file
+		*
+		* @return 1 if successfully moved to next file 0 if no next file exists
+		*/
+		short MoveToNextFile();
+		/**
+		* Function to move to the beginnning of the directory
+		* 
+		* @return 1 if successful, 0 otherwise
+		*/
+		short MoveToBeginning();
+		/**
+		* Function to check if the folder exists
+		* 
+		* @return whether or not the folder exists.
+		*/
+		bool GetFolderExists();
+		/**
+		* Function to check if current handle is directory or file
+		*
+		* @return true if current is a file, false otherwise. 
+		*/
+		bool GetCurIsFile();
+		/**
+		* Function to enter the current directory
+		*
+		* @return 1 if the directory was successfully entered, 0 otherwise
+		*/
+		short EnterDir(OUT Folder*& folder);
+		/**
+		* Function to open the current file for reading and writing
+		*
+		* @param file - a pointer where the function will store the opened file
+		* 
+		* @return 1 if the function was successful, 0 otherwise
+		*/
+		short Open(OUT File*& file);
+		/**
+		* Function to add a file to the directory
+		* 
+		* @param fileName - the name of the file
+		* @param file - a pointer to store the created file
+		*
+		* @return 1 if the function is successful or file already exists, 0 otherwise
+		*/
+		short AddFile(IN LPCWSTR fileName, OUT File*& file);
+		/**
+		* Function to remove current file and move to next handle
+		*
+		* @return 1 if the file was removed, 0 otherwise
+		* TODO: Add support for deleting folders
+		*/
+		short RemoveFile();
+		/**
+		* Function to return all files matching some attributes 
+		*
+		* @param attribs - the attributes for returned files to match, NULL gets everything
+		* @param recurse - whether or not to recursively search subdirectories
+		* @param recurDepth - the depth to recursively search, -1 recurses infinitely 
+		*
+		* @return all files that match the given parameters
+		*/
+		std::vector<File*>* GetFiles(IN FileSearchAttribs* attribs = NULL, IN int recurDepth = 0);
```

Comment:
Can we make this return a std::vector<File>? Or if that doesn't work, how about a std::shared_pointer<std::vector<std::shared_pointer<File>>>? Try to avoid normal pointers where at all possible